endpoint: avoid deferring code which locks endpoint in regenerateBPF

ianvernon opened the pull request:
If a panic occurs while this function is running and the endpoint lock is held
when the panic happens, the deferred proxy finalization / reverting code will
try to acquire the endpoint lock, resulting in the agent to hang. Now, the
function is returned to the caller of regenerateBPF to call when it returns.
This allows for a 'deferring' of the code in a sense, but the returned function
will only be called if a panic has not occurred.

More information about panic: https://blog.golang.org/defer-panic-and-recover

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6210)
<!-- Reviewable:end -->

Commit:
endpoint: avoid deferring code which locks endpoint in regenerateBPF

If a panic occurs while this function is running and the endpoint lock is held
when the panic happens, the deferred proxy finalization / reverting code will
try to acquire the endpoint lock, resulting in the agent to hang. Now, the
function is returned to the caller of regenerateBPF to call when it returns.
This allows for a 'deferring' of the code in a sense, but the returned function
will only be called if a panic has not occurred.

More information about panic: https://blog.golang.org/defer-panic-and-recover

Signed-off by: Ian Vernon <ian@cilium.io>

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -466,7 +466,9 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
 // boolean if the BPF compilation was executed and an error in case of an error.
-func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenContext *RegenerationContext) (revnum uint64, compiled bool, reterr error) {
+// Also returns a function which performs additional operations depending on
+// the state of the returned error.
```
The comment must say that the function must be invoked when it is non-nil.

ianvernon replied:
Done.

aanm reviewed (changes requested):
Unfortunately it won't be this simple, although the "`afterFunc`" is not executed in case of a panic, the caller of `regenerateBPF` has a `defer` function which also does a endpoint lock. So we will have the same problem.

Although I don't like it, I don't see a better way to do it than actually panic the whole program, if a panic happens we will panic hard and we need to fix why the panic happen which it is better than having the program in an unknown state. The way I suggest to fix this is to create a helper that can/ or probably should be used at the begging of all defer functions that does the following:

```golang
func PanicIfPanic() {
		if r := recover(); r != nil {
			panic(r)
		}
}
```
```golang
defer func() {
		PanicIfPanic()
		// lock endpoint ....
}
```

This way we won't deadlock because we don't stop the panic while doing a panic

Commit:
endpoint: move defer functions which lock endpoint to callers

Factor out logging of endpoint regeneration statistics into a new function which
is called after endpoint.regenerate().

Also add a new type, regenerationResult. This type contains a list of functions
to run in reverse order by the caller of endpoint.regenerate, and an error which
is populated if regeneration occurs. This ensures that no deadlock occurs in the
agent if a panic occurs during endpoint regeneration while the endpoint lock is
held, and a function has been deferred in the same callstack in which the panic
has occurred tries to acquire the endpoint lock.

Signed-off by: Ian Vernon <ian@cilium.io
